File: CapitalsQuiz.java
// CapitalsQuiz.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import com.google.gson.*;

public class CapitalsQuiz {
    private static final Map<String, String> COUNTRIES = new LinkedHashMap<>();
    private static final String STATS_FILE = "capitals_stats.json";
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    static {
        COUNTRIES.put("France", "Paris");
        COUNTRIES.put("Spain", "Madrid");
        COUNTRIES.put("Italy", "Rome");
        COUNTRIES.put("Germany", "Berlin");
        COUNTRIES.put("United Kingdom", "London");
        COUNTRIES.put("Portugal", "Lisbon");
        COUNTRIES.put("Netherlands", "Amsterdam");
        COUNTRIES.put("Belgium", "Brussels");
        COUNTRIES.put("Switzerland", "Bern");
        COUNTRIES.put("Austria", "Vienna");
        COUNTRIES.put("Greece", "Athens");
        COUNTRIES.put("Turkey", "Ankara");
        COUNTRIES.put("Russia", "Moscow");
        COUNTRIES.put("Ukraine", "Kyiv");
        COUNTRIES.put("Poland", "Warsaw");
        COUNTRIES.put("Sweden", "Stockholm");
        COUNTRIES.put("Norway", "Oslo");
        COUNTRIES.put("Denmark", "Copenhagen");
        COUNTRIES.put("Finland", "Helsinki");
        COUNTRIES.put("Ireland", "Dublin");
        COUNTRIES.put("USA", "Washington");
        COUNTRIES.put("Canada", "Ottawa");
        COUNTRIES.put("Mexico", "Mexico City");
        COUNTRIES.put("Brazil", "Brasilia");
        COUNTRIES.put("Argentina", "Buenos Aires");
        COUNTRIES.put("Chile", "Santiago");
        COUNTRIES.put("Peru", "Lima");
        COUNTRIES.put("Colombia", "Bogota");
        COUNTRIES.put("Venezuela", "Caracas");
        COUNTRIES.put("Australia", "Canberra");
        COUNTRIES.put("New Zealand", "Wellington");
        COUNTRIES.put("China", "Beijing");
        COUNTRIES.put("Japan", "Tokyo");
        COUNTRIES.put("South Korea", "Seoul");
        COUNTRIES.put("India", "New Delhi");
        COUNTRIES.put("Egypt", "Cairo");
        COUNTRIES.put("South Africa", "Pretoria");
        COUNTRIES.put("Nigeria", "Abuja");
        COUNTRIES.put("Kenya", "Nairobi");
    }

    static class Stats {
        int correct = 0;
        int incorrect = 0;
        int total = 0;
    }

    private Stats stats = new Stats();

    public CapitalsQuiz() {
        loadStats();
    }

    private void loadStats() {
        try {
            Path path = Paths.get(STATS_FILE);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                stats = gson.fromJson(json, Stats.class);
            }
        } catch (IOException e) {}
    }

    private void saveStats() {
        try {
            Files.write(Paths.get(STATS_FILE), gson.toJson(stats).getBytes());
        } catch (IOException e) {}
    }

    private String getHint(String capital, int level) {
        switch (level) {
            case 1:
                return "First letter is '" + capital.charAt(0) + "'";
            case 2:
                return "Has " + capital.length() + " letters";
            case 3:
                List<String> wrong = new ArrayList<>();
                for (String c : COUNTRIES.values()) {
                    if (!c.equals(capital) && wrong.size() < 2) {
                        wrong.add(c);
                    }
                }
                List<String> options = new ArrayList<>(wrong);
                options.add(capital);
                Collections.shuffle(options);
                return "Choose one: " + String.join(", ", options);
            default:
                return "";
        }
    }

    private String readLine() {
        try {
            return new BufferedReader(new InputStreamReader(System.in)).readLine();
        } catch (IOException e) {
            return "";
        }
    }

    public void runQuiz(int rounds) {
        System.out.println("\n🌍 Capitals Quiz");
        System.out.println("Round 1/" + rounds + "\n");

        List<String> countryList = new ArrayList<>(COUNTRIES.keySet());
        Collections.shuffle(countryList);
        List<String> selected = countryList.subList(0, Math.min(rounds, countryList.size()));

        int correct = 0, incorrect = 0;

        for (int i = 0; i < selected.size(); i++) {
            String country = selected.get(i);
            String capital = COUNTRIES.get(country);
            int hintLevel = 0;
            System.out.println("Country: " + country);
            while (true) {
                System.out.print("Your answer (or 'hint', 'skip', 'quit'): ");
                String input = readLine().trim().toLowerCase();
                switch (input) {
                    case "quit":
                        System.out.println("Quitting...");
                        return;
                    case "skip":
                        System.out.println("Skipped. The capital is " + capital);
                        incorrect++;
                        break;
                    case "hint":
                        hintLevel++;
                        if (hintLevel > 3) {
                            System.out.println("No more hints available.");
                        } else {
                            System.out.println("💡 Hint: " + getHint(capital, hintLevel));
                        }
                        continue;
                    default:
                        if (input.equals(capital.toLowerCase())) {
                            System.out.println("✅ Correct!");
                            correct++;
                            break;
                        } else {
                            System.out.println("❌ Wrong. Try again or type 'skip'.");
                            continue;
                        }
                }
                break;
            }
            System.out.printf("Score: %d/%d (%.1f%%)\n\n", correct, i+1, (correct/(double)(i+1)*100));
        }

        stats.correct += correct;
        stats.incorrect += incorrect;
        stats.total += correct + incorrect;
        saveStats();
        System.out.printf("Quiz finished! Correct: %d, Incorrect: %d\n", correct, incorrect);
    }

    public void showStats() {
        System.out.println("\n📊 Statistics");
        System.out.println("Correct: " + stats.correct);
        System.out.println("Incorrect: " + stats.incorrect);
        System.out.println("Total: " + stats.total);
        if (stats.total > 0) {
            double pct = (double) stats.correct / stats.total * 100;
            System.out.printf("Accuracy: %.1f%%\n", pct);
        }
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: java CapitalsQuiz [start|stats|help]");
            return;
        }
        CapitalsQuiz quiz = new CapitalsQuiz();
        String cmd = args[0].toLowerCase();
        switch (cmd) {
            case "start":
                quiz.runQuiz(10);
                break;
            case "stats":
                quiz.showStats();
                break;
            case "help":
                System.out.println("Commands: start, stats, help");
                break;
            default:
                System.out.println("Unknown command");
        }
    }
}

File: CapitalsQuiz.cs
// CapitalsQuiz.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class Stats
{
    [JsonPropertyName("correct")]
    public int Correct { get; set; }
    [JsonPropertyName("incorrect")]
    public int Incorrect { get; set; }
    [JsonPropertyName("total")]
    public int Total { get; set; }
}

class CapitalsQuiz
{
    private static readonly Dictionary<string, string> Countries = new Dictionary<string, string>
    {
        {"France", "Paris"}, {"Spain", "Madrid"}, {"Italy", "Rome"},
        {"Germany", "Berlin"}, {"United Kingdom", "London"},
        {"Portugal", "Lisbon"}, {"Netherlands", "Amsterdam"},
        {"Belgium", "Brussels"}, {"Switzerland", "Bern"},
        {"Austria", "Vienna"}, {"Greece", "Athens"},
        {"Turkey", "Ankara"}, {"Russia", "Moscow"},
        {"Ukraine", "Kyiv"}, {"Poland", "Warsaw"},
        {"Sweden", "Stockholm"}, {"Norway", "Oslo"},
        {"Denmark", "Copenhagen"}, {"Finland", "Helsinki"},
        {"Ireland", "Dublin"}, {"USA", "Washington"},
        {"Canada", "Ottawa"}, {"Mexico", "Mexico City"},
        {"Brazil", "Brasilia"}, {"Argentina", "Buenos Aires"},
        {"Chile", "Santiago"}, {"Peru", "Lima"},
        {"Colombia", "Bogota"}, {"Venezuela", "Caracas"},
        {"Australia", "Canberra"}, {"New Zealand", "Wellington"},
        {"China", "Beijing"}, {"Japan", "Tokyo"},
        {"South Korea", "Seoul"}, {"India", "New Delhi"},
        {"Egypt", "Cairo"}, {"South Africa", "Pretoria"},
        {"Nigeria", "Abuja"}, {"Kenya", "Nairobi"}
    };

    private const string StatsFile = "capitals_stats.json";
    private static readonly JsonSerializerOptions Options = new JsonSerializerOptions { WriteIndented = true };

    private Stats stats = new Stats();

    public CapitalsQuiz() => LoadStats();

    private void LoadStats()
    {
        if (!File.Exists(StatsFile)) return;
        string json = File.ReadAllText(StatsFile);
        stats = JsonSerializer.Deserialize<Stats>(json) ?? new Stats();
    }

    private void SaveStats()
    {
        string json = JsonSerializer.Serialize(stats, Options);
        File.WriteAllText(StatsFile, json);
    }

    private string GetHint(string capital, int level)
    {
        switch (level)
        {
            case 1: return $"First letter is '{capital[0]}'";
            case 2: return $"Has {capital.Length} letters";
            case 3:
                var wrong = Countries.Values.Where(c => c != capital).Take(2).ToList();
                var options = wrong.Append(capital).OrderBy(x => Guid.NewGuid()).ToList();
                return $"Choose one: {string.Join(", ", options)}";
            default: return "";
        }
    }

    public void RunQuiz(int rounds)
    {
        Console.WriteLine("\n🌍 Capitals Quiz");
        Console.WriteLine($"Round 1/{rounds}\n");

        var countryList = Countries.Keys.ToList();
        countryList = countryList.OrderBy(x => Guid.NewGuid()).ToList();
        var selected = countryList.Take(rounds).ToList();

        int correct = 0, incorrect = 0;

        for (int i = 0; i < selected.Count; i++)
        {
            string country = selected[i];
            string capital = Countries[country];
            int hintLevel = 0;
            Console.WriteLine($"Country: {country}");
            while (true)
            {
                Console.Write("Your answer (or 'hint', 'skip', 'quit'): ");
                string input = Console.ReadLine()?.Trim().ToLower() ?? "";
                switch (input)
                {
                    case "quit":
                        Console.WriteLine("Quitting...");
                        return;
                    case "skip":
                        Console.WriteLine($"Skipped. The capital is {capital}");
                        incorrect++;
                        break;
                    case "hint":
                        hintLevel++;
                        if (hintLevel > 3)
                            Console.WriteLine("No more hints available.");
                        else
                            Console.WriteLine($"💡 Hint: {GetHint(capital, hintLevel)}");
                        continue;
                    default:
                        if (input == capital.ToLower())
                        {
                            Console.WriteLine("✅ Correct!");
                            correct++;
                            break;
                        }
                        else
                        {
                            Console.WriteLine("❌ Wrong. Try again or type 'skip'.");
                            continue;
                        }
                }
                break;
            }
            Console.WriteLine($"Score: {correct}/{i+1} ({correct/(double)(i+1)*100:F1}%)\n");
        }

        stats.Correct += correct;
        stats.Incorrect += incorrect;
        stats.Total += correct + incorrect;
        SaveStats();
        Console.WriteLine($"Quiz finished! Correct: {correct}, Incorrect: {incorrect}");
    }

    public void ShowStats()
    {
        Console.WriteLine("\n📊 Statistics");
        Console.WriteLine($"Correct: {stats.Correct}");
        Console.WriteLine($"Incorrect: {stats.Incorrect}");
        Console.WriteLine($"Total: {stats.Total}");
        if (stats.Total > 0)
        {
            double pct = (double)stats.Correct / stats.Total * 100;
            Console.WriteLine($"Accuracy: {pct:F1}%");
        }
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: dotnet run -- [start|stats|help]");
            return;
        }
        var quiz = new CapitalsQuiz();
        string cmd = args[0].ToLower();
        switch (cmd)
        {
            case "start":
                quiz.RunQuiz(10);
                break;
            case "stats":
                quiz.ShowStats();
                break;
            case "help":
                Console.WriteLine("Commands: start, stats, help");
                break;
            default:
                Console.WriteLine("Unknown command");
                break;
        }
    }
}
